monitor/monitor.go

```diff
@@ -193,6 +194,17 @@ func (ml *monitorListener) drainQueue() {
 		if _, err := ml.conn.Write(msgBuf); err != nil {
 			ml.conn.Close()
 			ml.remove()
+
+			if op, ok := err.(*net.OpError); ok {
+				if syscerr, ok := op.Err.(*os.SyscallError); ok {
+					if errn, ok := syscerr.Err.(syscall.Errno); ok {
+						if errn == syscall.EPIPE {
```

Comment:
I'm not sure if this is the only type of error we would want to turn into an info. There are a few situations (a send timeout, a full send buffer) where a client is likely gone and we would end up here. The problem is that we don't define an API or mechanism for a client to say kthx done. It's only done via a disconnect. In a similar vein, the node-monitor (the agent) isn't broken when this happens, it's a purely housekeeping operation from its perspective (for better or worse).

I wonder if it would be simplest to just make the Warn below an info (but include the error anyway). It then matches the "client" connected Info, still shows us the error in case something bad is happening, and we avoid this 3 layer cast. I'm really surprised there isn't a errToErrno call somewhere!

Reply:
@raybejjani I could not reproduce any other error on monitor daemon side while termintating `cilium monitor`, that's why I am checking specifically for EPIPE. If you know other errors that shouldn't be logged, can you list them? I can add them to this patch.

Your proposed way is easier, but I think error message in logs is not needed here, in "normal" monitor usage. Showing error suggests that something is wrong here, but it isnt't (which is the main point in original issue).

But still we want to see error message when something is completely out of place. That's why I left the original warning if error returned is not something we think about.

PS. I also think that 3 layer casting is ridiculous way to retrieve Errno, but I could not find a more concise way. I can move it to helper function.

Reply:
That's my point, any network error is actually ok, and the monitor will disconnect the client as part of normal operation. The Warn is too strong, and an Info matches the Info we have when the client connects. I just realised that this is a unix socket, probably, so there is less room for networking type errors (like the peer can't just stop responding because if the other program closes the kernel will close its side of the connection, I'm pretty sure).

I can't think of other errors on a write of this sort but why would they be any different? the connection has been established, and those errors handled if they happened. Any error on the write call is basically saying: I can't send the data and we use that as a client remove signal (the client would reconnect if it wanted to). In that sense, printing all errors, including the EPIPE, is just the reason for removing that client.

I realise this is a bit contrived. The problem is that the "API" here is ill defined. I'm also probably a bit put off by the triple cast (I also looked for a way to do it and couldn't find anything simpler). A helper function is better, but to be an Info with an error message is less bad than the new code.

Anyway, I'll ask the ticket author what the real problem was, and that will decide it for us:
- if the problem was that people are afraid of Warns, we can choose to just make it an INFO
- if the problem was that people are afraid of the error text then your proposal is the only way to fix that.

Reply:
@tgraf can you tell if we want to get rid of the warn in all cases and leave the error there (I think it would make grepping logs by `error` string harder), or should we check errno from write error and log only errors that were not known to happen when disconnecting monitors?